trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/remote/RemoteSessionController.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.s3.proxy.server.remote;
+
+import com.google.inject.Inject;
+import io.trino.s3.proxy.server.credentials.Credential;
+import io.trino.s3.proxy.server.credentials.Credentials;
+import io.trino.s3.proxy.server.credentials.CredentialsController;
+import jakarta.annotation.PreDestroy;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.endpoints.Endpoint;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.sts.StsClient;
+import software.amazon.awssdk.services.sts.auth.StsAssumeRoleCredentialsProvider;
+
+import java.io.Closeable;
+import java.time.Instant;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class RemoteSessionController
```

Comment:
i like this interop between the `SigningController` and the `CredentialsController`, but I think the name `RemoteSessionController` is a bit misleading now. 


It's only public function is `withCredentials`, so maybe this should now be called the `CredentialController` and and the current credentials controller interface can be called `CredentialsProvider` (since that is what its really doing).


On the same note, the current `CredentialsController` interface has the `assumeRole` function - this now feels better suited to this class or a separate interface.

Part of my rationale is that the current `CredentialsProvider` will be an important plugin interface, so I think its worth keeping it light.

Reply:
I'll ping you in Slack for a more detailed conversation...

Reply:
Discussed in Slack - changes made.